app/lib/expression_viz_service.rb

```diff
@@ -137,6 +137,44 @@ def self.load_expression_data_array_points(study, gene, cluster, annotation, sub
     viz_data
   end
 
+  def self.load_correlated_data_array_scatter(study, genes, cluster, annotation,  subsample_threshold=nil)
+    viz_data = ClusterVizService.load_cluster_group_data_array_points(study, cluster, annotation, subsample_threshold=nil, include_coords: false)
+
+    gene0_expression = viz_data[:cells].map { |cell| genes[0]['scores'][cell].to_f.round(4) }
+    gene1_expression = viz_data[:cells].map { |cell| genes[1]['scores'][cell].to_f.round(4) }
+
+    viz_data[:x] = gene0_expression
+    viz_data[:y] = gene1_expression
+    viz_data[:pearsonsR] = pearsons_r(gene0_expression, gene1_expression).to_f.round(4)
+    viz_data
+  end
+
+  # adapted from https://blog.chrislowis.co.uk/2008/11/24/ruby-gsl-pearson.html
+  # note that the age of the blog means his performance numbers of the ruby version are off
+  # in local testing, this can compute correlation for 1MM cells in 0.5 seconds
+  def self.pearsons_r(x,y)
```

Comment:
Having a hand-transcribed algorithm like this is pretty neat, but also makes me wary.  Ideally, we'd use a Ruby statistics library that's been tested, gets bug reports if something is wrong, etc.

Are any of these suitable?

* https://github.com/sciruby/statsample
* https://github.com/clbustos/statsample
* https://github.com/alfonsojimenez/pearson

Reply:
I totally concur that this should be in a gem if possible.  The problem was my 15 minutes of googling turned up nothing.   I'd looked at the statsample gems, and none had recent updates (e.g. sciruby statsample was last updated in 2017), and  both have complex dependencies on native C extensions that suggested it would not be worth bringing in for a single function.  I hadn't seen the alfonsojimenez/pearson gem.  But as best as I can tell that's not maintained or widely used.

It boggles my mind that there isn't a standard, up-to-date Ruby stats library.  I guess all the cool kids really do use Python

Reply:
Update -- after chatting with Sarah, she suggested using both spearman and Pearson, which means we should definitely start using a gem, and we can deal with the added install complexity

Reply:
double update, the gem is performance prohibitive.  So we'll try client side, and start with just Spearman, since that's the most 'correct'